### ISSUE TITLE
Broken in elixir 1.10

### ISSUE THREAD

**Issue by austinsmorris** [Nebo15/logger_json#31]
Please see #30 and the travis test for elixir 1.10.

**Comment by karlseguin** [comment]
As a workaround, you can use:
```
backend: [metadata: [:file, :line, :function, :module, :application]]
```

instead of 
```
backend: [metadata: :all]
```

**Comment by hauleth** [comment]
Just to mention, this is bug in the `LoggerJson` not in Elixir. Everything is due to fact that `LoggerJson` always assumed that all values in metadata will have implemented `Jason.Encode` protocol which was never a case (metadata is defined as `keyword()` which allows any term as a value). Elixir 1.10 just adds new values possible in metadata, and these are PIDs, tuples, and functions (as used in Erlang Logger) which do not have "canonical" JSON representation, which mean that there is no default `Jason.Encoder` defined. The quick fix would be to use:

```elixir
def encode(term) do
  case Jason.Encode.impl_for(term) do
    nil -> inspect(term)
    module -> module.encode(term, [])
  end
end
```

**Comment by timrogers** [comment]
> As a workaround, you can use:
> 
> ```
> backend: [metadata: [:file, :line, :function, :module, :application]]
> ```
> 
> instead of
> 
> ```
> backend: [metadata: :all]
> ```

If you use your own custom logger metadata, this workaround won't work (or at least, you'll have to add all of the custom keys you use to the whitelist).

**Comment by AndrewDryga** [comment]
@hauleth I wish it would be that easy but the problem is that we would need to encode it recursively or give up on an ability to have nested keyword/maps inside of metadata (which I would avoid). Another option is to implement `Jason.Encoder` protocol for all of them but it would affect the application and may result in conflicts.

**Comment by hauleth** [comment]
@AndrewDryga maybe something like:

```elixir
def encode(list) when is_list(list),
  do: Enum.map(list, &encode/1)

def encode(map) when not is_struct(map) do
  safe_map =
    for {key, value} <- map,
      into: %{},
      do: {encode(key), encode(value)}

  Json.encode!(safe_map)
end

def encode(term) do
  case Jason.Encode.impl_for(term) do
    nil -> inspect(term)
    module -> module.encode(term, [])
  end
end
```

**Comment by jbernardo95** [comment]
What if we simply ignore the keys that can't be encoded? I opened a PR to do this [here](https://github.com/Nebo15/logger_json/pull/33). Let me know what you think!

**Comment by austinsmorris** [comment]
This issue is closed by #33.  Just waiting on a new release to hit hex.  Thanks all!

**Comment by AndrewDryga** [comment]
3.3.0 is out, please let me know if it breaks again. For now, I'll close this issue.

I want to thank everyone for your contribution and thank you for using this package 🎊.
